internal/cephfs/util/mountinfo.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	// google.golang.org/protobuf/encoding doesn't offer MessageV2().
+	"github.com/golang/protobuf/proto" // nolint:staticcheck // See comment above.
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+// This file provides functionality to store various mount information
+// in a file. It's currently used to restore ceph-fuse mounts.
+// Mount info is stored in `/csi/mountinfo`.
+
+const (
+	mountinfoDir = "/csi/mountinfo"
+)
+
+// nodeStageMountinfoRecord describes a single
+// record of mountinfo of a staged volume.
+// encoding/json-friendly format.
+// Only for internal use for marshaling and unmarshaling.
+type nodeStageMountinfoRecord struct {
+	VolumeCapabilityProtoJSON string            `json:",omitempty"`
+	MountOptions              []string          `json:",omitempty"`
+	Secrets                   map[string]string `json:",omitempty"`
```

Comment:
why store the secrets on the nodePath what is the problem with getting it from the Nodepublish secret?

Reply:
Please see the comments in https://github.com/ceph/ceph-csi/pull/2634#pullrequestreview-894655827. To sum this up:
* Users would need to recreate their storage classes after upgrade to take advantage of this feature.
* The same is true for existing PVs that would need to be recreated. This would mean taking the application using these PVs down.
* Storing the keyrings on the node doesn't seem to be an issue. Should an adversary user get root access to the node, it is trivial to acquire the keys anyway. See https://github.com/ceph/ceph-csi/issues/2924 for more details. Instead of having this "false feeling of safety", we could focus on limiting the "blast radius" a cephx user of a volume has. But this is out of scope for this PR.

Reply:
as the fuse plugin is not production-ready and it's a new storageclass for the fuse mounter type why worry about adding new publish secrets to the storageclass.

IMO we don't need to worry about supporting the existing PVs.

Reply:
am not sure what else major problem this can lead to and one this I can think of now itself is that this will also block us from updating the secret details username/keyring value. am trying to avoid a bigger problem when we GA this one.

One use case to save the secrets is to support already created PVC. we can document users to delete and recreate the PV object with publish reference also. IMO no one is using cephfs fuse in production yet.

Reply:
> and one this I can think of now itself is that this will also block us from updating the secret details username/keyring value.

I've touched on this in my last comment.

>  Things like credentials revocation (so that the mountinfo file would possibly contain stale credentials) is also of no concern, because you'd still have to fully unmount the volume from the node, so that NodeUnstageVolume is triggered

If you change client name or the key, you need to remount. Remounting will refresh the mountinfo file.

> IMO no one is using cephfs fuse in production yet.

We are, in many clusters.

Reply:
> am trying to avoid a bigger problem when we GA this one.

Actually adding new secrets is an API-facing breaking change, whose necessity might change in the future. Right now we can have this feature without needing to change the API (the secrets requirements in this case). Should the needs change, we can always add them later... it's always much harder to remove things though.